services/federator/test/integration/Test/Federator/InwardSpec.hs

```diff
@@ -73,24 +76,70 @@ spec env =
-          GRpcOk (InwardResponseErr err) -> assertFailure $ "Unexpected error response: " <> show err
-          x -> assertFailure $ "GRpc call failed unexpectedly: " <> show x
-
--- All the code below is copied from brig-integrration tests
+    it "should give InvalidEndpoint on a 404 'no-endpoint'" $
+      runTestFederator env $ do
+        err <- asInwardError =<< inwardBrigCall "federation/this-endpoint-does-not-exist" (encode Aeson.emptyObject)
+        expectErr IInvalidEndpoint err
+
+-- Utility functions
+--
+expectErr :: InwardErrorType -> InwardError -> TestFederator IO ()
+expectErr expectedType err = do
+  case inwardErrorType err of
+    t | t == expectedType -> pure ()
+    _ -> liftIO $ assertFailure $ "expected type '" <> show expectedType <> "' but got " <> show err
+
+asInwardBody :: forall a. (HasCallStack, Typeable a, FromJSON a) => GRpcReply InwardResponse -> TestFederator IO a
+asInwardBody = either err pure . asInwardBodyEither
+  where
+    err parserErr = liftIO $ assertFailure parserErr
+
+asInwardBodyUnsafe :: (HasCallStack, Typeable a, FromJSON a) => GRpcReply InwardResponse -> a
+asInwardBodyUnsafe = either err id . asInwardBodyEither
+  where
+    err parserErr = error . unwords $ ["asInwardBodyUnsafe:"] <> [parserErr]
+
+asInwardBodyEither :: forall a. (HasCallStack, Typeable a, FromJSON a) => GRpcReply InwardResponse -> Either String a
+asInwardBodyEither (GRpcOk (InwardResponseError err)) = Left (show err)
+asInwardBodyEither (GRpcOk (InwardResponseBody bdy)) = first addTypeInfo $ eitherDecodeStrict bdy
+  where
+    addTypeInfo :: String -> String
+    addTypeInfo = (("Could not parse InwardResponseBody as '" <> show (typeRep (Data.Proxy.Proxy @a)) <> "': ") <>)
+asInwardBodyEither other = Left $ "GRpc call failed unexpectedly: " <> show other
+
+asInwardError :: HasCallStack => GRpcReply InwardResponse -> TestFederator IO InwardError
+asInwardError = either err pure . asInwardErrorEither
+  where
+    err parserErr = liftIO $ assertFailure (unwords $ ["asInwardError:"] <> [parserErr])
+
+asInwardErrorUnsafe :: HasCallStack => GRpcReply InwardResponse -> InwardError
+asInwardErrorUnsafe = either err id . asInwardErrorEither
+  where
+    err parserErr = error . unwords $ ["asInwardErrorUnsafe:"] <> [parserErr]
+
+asInwardErrorEither :: HasCallStack => GRpcReply InwardResponse -> Either String InwardError
+asInwardErrorEither (GRpcOk (InwardResponseError err)) = Right err
+asInwardErrorEither (GRpcOk (InwardResponseBody bdy)) = Left ("expected InwardError, but got InwardResponseBody: " <> show bdy)
+asInwardErrorEither other = Left $ "GRpc call failed unexpectedly: " <> show other
+
+inwardBrigCall :: (MonadIO m, MonadHttp m, MonadReader TestEnv m, HasCallStack) => ByteString -> LBS.ByteString -> m (GRpcReply InwardResponse)
+inwardBrigCall requestPath payload = do
+  c <- viewFederatorExternalClient
+  let brigCall = Request Brig requestPath (LBS.toStrict payload) "foo.example.com"
```

Comment:
What is this `"foo.example.com"` argument? Would it make sense to add a note?

Reply:
It's the originDomain, which is defined as part of Request data type, here: https://github.com/wireapp/wire-server/blob/develop/libs/wire-api-federation/src/Wire/API/Federation/GRPC/Types.hs#L135-L140. Maybe the object can be constructed using its names for clarity, good idea.